packages/backend/server/src/core/config/service.ts

```diff
@@ -1,17 +1,121 @@
-import { Injectable } from '@nestjs/common';
-import { PrismaClient } from '@prisma/client';
+import { Injectable, Logger, OnApplicationBootstrap } from '@nestjs/common';
+import { Cron, CronExpression } from '@nestjs/schedule';
+import { set } from 'lodash-es';
+
+import { ConfigFactory, EventBus, OnEvent } from '../../base';
+import { Models } from '../../models';
+import { ServerFeature } from './types';
+
+declare global {
+  interface Events {
+    'config.init': {
+      config: DeepReadonly<AppConfig>;
+    };
+    'config.changed': {
+      updates: DeepPartial<AppConfig>;
+    };
+    'config.changed.broadcast': {
+      updates: DeepPartial<AppConfig>;
+    };
+  }
+}
 
 @Injectable()
-export class ServerService {
+export class ServerService implements OnApplicationBootstrap {
   private _initialized: boolean | null = null;
-  constructor(private readonly db: PrismaClient) {}
+  readonly #features = new Set<ServerFeature>();
+  readonly #logger = new Logger(ServerService.name);
+
+  constructor(
+    private readonly models: Models,
+    private readonly configFactory: ConfigFactory,
+    private readonly event: EventBus
+  ) {}
+
+  async onApplicationBootstrap() {
+    await this.setup();
+  }
+
+  get features() {
+    return Array.from(this.#features);
+  }
 
   async initialized() {
     if (!this._initialized) {
-      const userCount = await this.db.user.count();
+      const userCount = await this.models.user.count();
       this._initialized = userCount > 0;
     }
 
     return this._initialized;
   }
+
+  enableFeature(feature: ServerFeature) {
+    this.#features.add(feature);
+  }
+
+  disableFeature(feature: ServerFeature) {
+    this.#features.delete(feature);
+  }
+
+  get config() {
+    return this.configFactory.config;
+  }
+
+  async updateConfig(
+    user: string,
+    updates: Array<{ module: string; key: string; value: any }>
+  ) {
+    this.configFactory.validate(updates);
+
+    const promises = await this.models.appConfig.save(
+      user,
+      updates.map(update => ({
+        key: `${update.module}.${update.key}`,
+        value: update.value,
+      }))
+    );
+
+    const overrides: DeepPartial<AppConfig> = {};
+    // only take successfully saved configs
+    promises.forEach(promise => {
+      if (promise.status === 'fulfilled') {
+        set(overrides, promise.value.id, promise.value.value);
+      } else {
+        this.#logger.error(`Failed to save app config`, promise.reason);
+      }
+    });
+    this.configFactory.override(overrides);
+    this.event.emit('config.changed', { updates: overrides });
+    this.event.broadcast('config.changed.broadcast', { updates: overrides });
+  }
+
+  @OnEvent('config.changed.broadcast')
+  onConfigChangedBroadcast(updates: DeepPartial<AppConfig>) {
+    this.configFactory.override(updates);
+    this.event.emit('config.changed', { updates });
+  }
+
+  @Cron(CronExpression.EVERY_5_MINUTES)
```

Comment:
Let's not add this scheduled full update for now, as it will trigger a full config.changed event every 5 minutes, causing many clients to reinitialize. 
For instance, smtp will recreate the tcp connection, and there are some clients for which we are unsure if repeated initialization might lead to memory leak issues.

Reply:
we need a diff here